syntax = "proto3";


message NewRatingRequest {
    int32 user_id = 1;
    int32 movie_id = 2;
    int32 rating = 3;
}

message RelevantMovieRequest {
    int32 user_id = 1;
    int32 offset = 2;
    int32 limit = 3;
}

message SimilarMovieRequest {
    int32 movie_id = 1;
    int32 offset = 2;
    int32 limit = 3;
}

message RelevantSimilarMovieRequest {
    int32 user_id = 1;
    int32 movie_id = 2;
    int32 offset = 3;
    int32 limit = 4;
}

message MovieRecommendation {
    int32 id = 1;
}

message EmptyResponse {
    bool ok = 1;
}

message RecommendationResponse {
    repeated MovieRecommendation recommendations = 1;
}

service Recommender {
    rpc SaveNewRating (NewRatingRequest) returns (EmptyResponse);
    rpc RecommendMovie (RelevantMovieRequest) returns (RecommendationResponse);
    rpc RecommendSimilarMovie (SimilarMovieRequest) returns (RecommendationResponse);
    rpc RecommendRelevantSimilarMovie (RelevantSimilarMovieRequest) returns (RecommendationResponse);
}
